hpx/util/bind_front.hpp

```diff
@@ -0,0 +1,253 @@
+//  Copyright (c) 2011-2012 Thomas Heller
+//  Copyright (c) 2013-2017 Agustin Berge
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_UTIL_BIND_FRONT_HPP
+#define HPX_UTIL_BIND_FRONT_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/traits/get_function_address.hpp>
+#include <hpx/traits/get_function_annotation.hpp>
+#include <hpx/util/decay.hpp>
+#include <hpx/util/detail/pack.hpp>
+#include <hpx/util/invoke.hpp>
+#include <hpx/util/result_of.hpp>
+#include <hpx/util/tuple.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace util
+{
+    namespace detail
+    {
+        ///////////////////////////////////////////////////////////////////////
+        template <typename F, typename Ts, typename ...Us>
+        struct invoke_bound_front_result;
+
+        template <typename F, typename ...Ts, typename ...Us>
+        struct invoke_bound_front_result<F&, util::tuple<Ts...>&, Us...>
+          : util::invoke_result<F&, Ts&..., Us...>
+        {};
+
+        template <typename F, typename ...Ts, typename ...Us>
+        struct invoke_bound_front_result<F const&, util::tuple<Ts...> const&, Us...>
+          : util::invoke_result<F const&, Ts const&..., Us...>
+        {};
+
+        template <typename F, typename ...Ts, typename ...Us>
+        struct invoke_bound_front_result<F&&, util::tuple<Ts...>&&, Us...>
+          : util::invoke_result<F, Ts..., Us...>
+        {};
+
+        template <typename F, typename ...Ts, typename ...Us>
+        struct invoke_bound_front_result<F const&&, util::tuple<Ts...> const&&, Us...>
+          : util::invoke_result<F const, Ts const..., Us...>
+        {};
+
+        template <typename F, std::size_t ...Is, typename Ts, typename ...Us>
+        HPX_HOST_DEVICE
+        typename invoke_bound_front_result<F&&, Ts&&, Us...>::type
+        bound_front_impl(F&& f, pack_c<std::size_t, Is...>, Ts&& bound,
+            Us&&... unbound)
+        {
+            return util::invoke(std::forward<F>(f),
+                util::get<Is>(std::forward<Ts>(bound))...,
```

Comment:
This might lead to 'use after move' if `bound` is a rvalue. This should read `std::forward<decltype(util::get<Is>(bound))>(util::get<Is>(bound))...` (or similar) instead.